source/extensions/stat_sinks/common/statsd/statsd.cc

```diff
@@ -61,14 +61,14 @@ void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
   Buffer::OwnedImpl buffer;
 
   for (const auto& counter : snapshot.counters()) {
-    if (counter.counter_.get().used()) {
+    if (counter.counter_.get().used() && !counter.counter_.get().hidden()) {
```

Comment:
nit: we don't have an API to hide counters right now., thought it is supported in the data model. Until it's possible to hit that branch, we might as well not pay the cost for every counter; there's a lot of counters.

Instead can you leave a note to this effect here but don't change this code?

However, I just realized we are in statsd.cc. I think what we want to do is change the AllocatorImpl::forEachSinkedGauge et al in  source/common/stats/allocator_impl.cc:419. source/server/server.cc:198. By doing this we are altering the definition of "sinked" to exclude hidden elements.

We can leave comments for the other 2 stat types handled by AllocatorImpl, but just change the code for Gauges.

I would *keep* the tests you added for the various stats sinks in addition to adding ones for allocator_impl_test.cc.

@pradeepcrao WDYT?

Reply:
I feel silly for not thinking of this earlier. Yes, that is a much better solution (doing this in `allocator_impl.cc`). This also makes it much easier to use `SinkPredicates` for this down the line if needed.

Reply:
@DiazAlan my sincerest apologies for the rework.

Reply:
Also let's not toss all the good work Alan did for sink-specific tests; that can stay.

Reply:
@jmarantz I take back what I said. I propose that we instead make the change in the creation of `MetricSnapshotImpl` https://github.com/envoyproxy/envoy/blob/main/source/server/server.cc#L215-L217

If `SinkPredicates` is not defined, we just call `forEach<Metric>` instead of `forEachSinked<Metric>`, which should be the case of most users of Envoy. 🤦 

WDYT?

@DiazAlan If we go with the approach I suggested here, we can test this by invoking `InstanceUtil::flushMetricsToSinks`.

Reply:
IIUC you are debating whether to implement the hidden-filtering in the single implementation of forEachSinkedGauge in allocator_impl.cc or its single call-site in MetricsSnapshotImpl.

Either work for me i guess, as the code stands, but I was chatting with @krajshiva about potentially adding more calls to forEachSinkedGauge because MetricsSnapshotImpl is really inefficient cause it has to create this huge vector of all counters. The sinks should just call the foreach loops directly, and that supports putting it at the source, in allocator_impl.cc.

You could also argue to put it in the Store instead: https://github.com/envoyproxy/envoy/blob/main/source/common/stats/thread_local_store.cc#L1059 . I slightly favor putting it in the allocator to that, for a trivial performance nit, but don't feel too strongly. 

The downside of putting it  in allocator_impl.cc is that it has to cover both the cases where a predicate is defined when there isn't, but that doesn't seem like a big deal to me.

Reply:
How do you feel about something like this? 

```
void AllocatorImpl::forEachSinkedGauge(SizeFn f_size, StatFn<Gauge> f_stat) const {
  auto f_stat_to_use = [&f_stat](Gauge& gauge) {if (!gauge.hidden()) f_stat(gauge); }; 
  if (sink_predicates_ != nullptr) {
    Thread::LockGuard lock(mutex_);
    f_size(sinked_gauges_.size());
    for (auto gauge : sinked_gauges_) {
      f_stat_to_use(*gauge);
    }
  } else {
    forEachGauge(f_size, f_stat_to_use);
  }
}
```

Reply:
sgtm.